### ISSUE TITLE
测试时为啥候选物品的embedding 不需要过Model模块的fc呢，也即MLP模块，而是直接使用物品的modalembedding或者说id embedding

### ISSUE THREAD

**Issue by Jackie-gj** [westlake-repl/IDvs.MoRec#7]
self.fc = MLP_Layers(word_embedding_dim=num_fc_ftr,
                             item_embedding_dim=args.embedding_dim,
                             layers=[args.embedding_dim] * (args.dnn_layer + 1),
                             drop_rate=args.drop_rate)这段代码在训练是对输入的embedding进行了转换，然后再与候选的正负样本计算相似度以及BCE损失， 在模型预测时，为啥是直接使用item_embeddings而不需要经过上面得MLP_Layers呢？
    item_embeddings = item_embeddings.to(local_rank)
    with torch.no_grad():
        eval_all_user = []
        item_rank = torch.Tensor(np.arange(item_num) + 1).to(local_rank)
        for data in eval_dl:
            user_ids, input_embs, log_mask, labels = data
            user_ids, input_embs, log_mask, labels = \
                user_ids.to(local_rank), input_embs.to(local_rank),\
                log_mask.to(local_rank), labels.to(local_rank).detach()
            prec_emb = model.module.user_encoder(input_embs, log_mask, local_rank)[:, -1].detach()
            scores = torch.matmul(prec_emb, item_embeddings.t()).squeeze(dim=-1).detach()

**Comment by zyuanat** [comment]
感谢提问，抱歉这么晚回复
所提到的item embedding其实是已经通过model模块的fc层的，在metics.py文件的get_item_embeddings()函数中的 “item_emb = model.module.fc(item_emb)”这一行：

`def get_item_embeddings(model, item_image_embs, test_batch_size, args, local_rank):
    model.eval()
    item_dataset = Build_Id_Eval_Dataset(data=item_image_embs)
    item_dataloader = DataLoader(item_dataset, batch_size=test_batch_size,
                                 num_workers=args.num_workers, pin_memory=True)
    item_embeddings = []
    with torch.no_grad():
        for input_ids in item_dataloader:
            item_emb = input_ids.to(local_rank)
            item_emb = model.module.fc(item_emb)
            item_embeddings.extend(item_emb)
    return torch.stack(tensors=item_embeddings, dim=0).to(torch.device("cpu")).detach()`

**Comment by Jackie-gj** [comment]
这个可加可不加的，这个模型应该没有用到



---- 回复的原邮件 ----
| 发件人 | ***@***.***> |
| 日期 | 2024年02月02日 21:59 |
| 收件人 | ***@***.***> |
| 抄送至 | ***@***.***>***@***.***> |
| 主题 | Re: [westlake-repl/IDvs.MoRec] 测试时为啥候选物品的embedding 不需要过Model模块的fc呢，也即MLP模块，而是直接使用物品的modalembedding或者说id embedding (Issue #7) |

感谢提问，抱歉这么晚回复
所提到的item embedding其实是已经通过model模块的fc层的，在metics.py文件的get_item_embeddings()函数中的 “item_emb = model.module.fc(item_emb)”这一行：

def get_item_embeddings(model, item_image_embs, test_batch_size, args, local_rank): model.eval() item_dataset = Build_Id_Eval_Dataset(data=item_image_embs) item_dataloader = DataLoader(item_dataset, batch_size=test_batch_size, num_workers=args.num_workers, pin_memory=True) item_embeddings = [] with torch.no_grad(): for input_ids in item_dataloader: item_emb = input_ids.to(local_rank) item_emb = model.module.fc(item_emb) item_embeddings.extend(item_emb) return torch.stack(tensors=item_embeddings, dim=0).to(torch.device("cpu")).detach()
